specs/DataSet.rst

```diff
@@ -0,0 +1,253 @@
+=====================
+DataSet Specification
+=====================
+
+Introduction
+============
+
+The DataSet class is used in QCoDeS to hold measurement results.
+It is the destination for measurement loops and the source for plotting and data analysis.
+As such, it is a central component of QCoDeS.
+
+The DataSet class should be usable on its own, without other QCoDeS components.
+In particular, the DataSet class should not require the use of Loop and parameters, although it should integrate with those components seamlessly.
+This will significantly improve the modularity of QCoDeS by allowing users to plug into and extend the package in many different ways.
+As long as a DataSet is used for data storage, users can freely select the QCoDeS components they want to use.
+
+Terminology
+================
+
+Parameter
+    A logically-single value input to or produced by a measurement.
+    A parameter need not be a scalar, but can be an array or a tuple or an array of tuples, etc.
+    A DataSet parameter corresponds conceptually to a QCoDeS parameter, but does not have to be defined by or associated with a QCoDeS Parameter . 
+    Roughly, a parameter represents a column in a table of experimental data.
+    
+Result
+    A result is the collection of parameter values associated to a single measurement in an experiment.
+    Roughly, a result corresponds to a row in a table of experimental data.
+    
+Role
+    Parameters may play different roles in a measurement.
+    Specifically, they may be input to the measurement (set-points) or outputs of the measurement (measured or computed values).
+    This distinction is important for plotting and for replicating an experiment.
```

Comment:
very much agree with this. From my experience the difficulty is the difference between the handling of 'real' setpoints (ie things that actually get set on the instrument) vs calculated setpoints (ie exactly the frequency example above which you dont directly set on the instrument or read from it but rather calculate from the values of some other parameters) vs actual measured values (returned by the instrument). I think some examples should be worked out before too much time is spent writing this aspect of the data set to avoid (as much as possible) problems coming up later which result in unpleasant hacking.